File: vb/VS2005/netframework/Modules/10.API/30.FlexCelImageExplorer/MainForm.vb
Imports System.Collections
Imports System.ComponentModel
Imports System.IO

Imports FlexCel.Core
Imports FlexCel.XlsAdapter

Namespace FlexCelImageExplorer
	''' <summary>
	''' Image Explorer.
	''' </summary>
	Partial Public Class mainForm
		Inherits System.Windows.Forms.Form

		Public Sub New()
			InitializeComponent()
			AddHandler GetCurrencyManager.CurrentChanged, AddressOf CurrentRowChanged
			ResizeToolbar(mainToolbar)
		End Sub

		Private Sub ResizeToolbar(ByVal toolbar As ToolStrip)

			Using gr As Graphics = CreateGraphics()
				Dim xFactor As Double = gr.DpiX / 96.0
				Dim yFactor As Double = gr.DpiY / 96.0
				toolbar.ImageScalingSize = New Size(CInt(Fix(24 * xFactor)), CInt(Fix(24 * yFactor)))
				toolbar.Width = 0 'force a recalc of the buttons.
			End Using
		End Sub

		Private CurrentFilename As String = Nothing
		Private CompressForm As TCompressForm

		Private ReadOnly Property GetCurrencyManager() As CurrencyManager
			Get
				Return CType(Me.BindingContext(dataGrid.DataSource, dataGrid.DataMember), CurrencyManager)
			End Get
		End Property

		Private ReadOnly Property GetImagePos() As Integer
			Get
				Return GetCurrencyManager.Position
			End Get
		End Property

		Private Sub btnExit_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnExit.Click
			Close()
		End Sub

		Private Sub btnOpenFile_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnOpenFile.Click
			If openFileDialog1.ShowDialog() <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If
			OpenFile(openFileDialog1.FileName)
			If cbScanFolder.Checked Then
				FillListBox()
			End If
		End Sub

		Private Function GetHasCrop(ByVal ImgProps As TImageProperties) As Boolean
			Return ImgProps.CropArea.CropFromLeft <> 0 OrElse ImgProps.CropArea.CropFromRight <> 0 OrElse ImgProps.CropArea.CropFromTop <> 0 OrElse ImgProps.CropArea.CropFromBottom <> 0
		End Function

		Private Sub FillListBox()
			lblFolder.Text = "Files on folder: " & Path.GetDirectoryName(openFileDialog1.FileName)
			Dim di As New DirectoryInfo(Path.GetDirectoryName(openFileDialog1.FileName))
			Dim Fi() As FileInfo = di.GetFiles("*.xls")
			FilesListBox.Items.Clear()

			Dim Files(Fi.Length - 1) As TImageInfo

			For k As Integer = 0 To Fi.Length - 1
				Dim f As FileInfo = Fi(k)
				Dim HasCrop As Boolean = False
				Dim HasARGB As Boolean = False
				Dim x1 As New XlsFile()

				Dim HasImages As Boolean = False

				Try
					x1.Open(f.FullName)
					For sheet As Integer = 1 To x1.SheetCount
						x1.ActiveSheet = sheet
						For i As Integer = x1.ImageCount To 1 Step -1
							HasImages = True
							Dim ip As TImageProperties = x1.GetImageProperties(i)
							If Not HasCrop Then
								HasCrop = GetHasCrop(ip)
							End If

							Dim imgType As TXlsImgType = TXlsImgType.Unknown
							Using ms As New MemoryStream()
								x1.GetImage(i, imgType, ms)
								Dim PngInfo As FlexCel.Pdf.TPngInformation = FlexCel.Pdf.TPdfPng.GetPngInfo(ms)
								If PngInfo IsNot Nothing Then
									HasARGB = PngInfo.ColorType = 6
								End If
							End Using

						Next i
					Next sheet
				Catch e1 As Exception
					Files(k) = New TImageInfo(f, False, False, False, False)
					Continue For
				End Try

				Files(k) = New TImageInfo(f, True, HasCrop, HasImages, HasARGB)
			Next k

			FilesListBox.Items.AddRange(Files)
		End Sub

		Private Sub OpenFile(ByVal FileName As String)
			ImageDataTable.Rows.Clear()

			Try
				Dim Xls As New XlsFile(True)
				CurrentFilename = FileName
				Xls.Open(FileName)

				For sheet As Integer = 1 To Xls.SheetCount
					Xls.ActiveSheet = sheet
					For i As Integer = Xls.ImageCount To 1 Step -1
						Dim ImageType As TXlsImgType = TXlsImgType.Unknown
						Dim ImgBytes() As Byte = Xls.GetImage(i, ImageType)
						Dim ImgProps As TImageProperties = Xls.GetImageProperties(i)
						Dim ImgData(ImageDataTable.Columns.Count - 1) As Object
						ImgData(0) = Xls.SheetName
						ImgData(1) = i
						ImgData(4) = ImageType.ToString()
						ImgData(7) = Xls.GetImageName(i)
						ImgData(8) = ImgBytes
						ImgData(9) = GetHasCrop(ImgProps)


						Using ms As New MemoryStream(ImgBytes)
							Dim PngInfo As FlexCel.Pdf.TPngInformation = FlexCel.Pdf.TPdfPng.GetPngInfo(ms)
							If PngInfo IsNot Nothing Then
								ImgData(2) = PngInfo.Width
								ImgData(3) = PngInfo.Height
								Dim s As String = String.Empty
								Dim bpp As Integer = 0

								If (PngInfo.ColorType And 4) <> 0 Then
									s &= "ALPHA-"
									bpp = 1
								End If
								If (PngInfo.ColorType And 2) = 0 Then
									s &= "Grayscale -" & (1 << PngInfo.BitDepth).ToString() & " shades. "
									bpp = 1
								Else
									If (PngInfo.ColorType And 1) = 0 Then
										bpp += 3
										s &= "RGB - " & (PngInfo.BitDepth * (bpp)).ToString() & "bpp.  "
									Else
										s &= "Indexed - " & (1 << PngInfo.BitDepth).ToString() & " colors. "
										bpp = 1
									End If
								End If

								ImgData(5) = s

								ImgData(6) = (Math.Round(PngInfo.Width * PngInfo.Height * PngInfo.BitDepth * bpp / 8F / 1024F)).ToString() & " kb."
							Else
								ms.Position = 0
								Try
									Using Img As Image = Image.FromStream(ms)
										Dim Bmp As Bitmap = TryCast(Img, Bitmap)
										If Bmp IsNot Nothing Then
											ImgData(5) = Bmp.PixelFormat.ToString() & "bpp"
										End If
										ImgData(2) = Img.Width
										ImgData(3) = Img.Height
									End Using
								Catch e1 As Exception
									ImgData(2) = -1
									ImgData(3) = -1
									ImgData(5) = Nothing
									ImgData(8) = Nothing

								End Try
							End If
						End Using


						ImageDataTable.Rows.Add(ImgData)
					Next i
				Next sheet

			Catch ex As Exception
				MessageBox.Show(ex.Message, "Error")
				dataGrid.CaptionText = "No file selected"
				CurrentFilename = Nothing
				Return
			End Try
			dataGrid.CaptionText = "Selected file: " & FileName
			CurrentRowChanged(GetCurrencyManager, Nothing)
		End Sub

		Private Sub FilesListBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles FilesListBox.SelectedIndexChanged
			Dim ImageInfo As TImageInfo = CType(FilesListBox.SelectedItem, TImageInfo)
			If ImageInfo Is Nothing Then
				Return
			End If
			OpenFile(ImageInfo.File.FullName)
		End Sub

		Public Sub CurrentRowChanged(ByVal sender As Object, ByVal e As System.EventArgs)
			Dim Pos As Integer = CType(sender, BindingManagerBase).Position
			If Pos < 0 Then
				PreviewBox.Image = Nothing
				Return
			End If

			Dim r As DataRowCollection = dataSet1.Tables("ImageDataTable").Rows
			If Pos < 0 OrElse Pos >= r.Count Then
				PreviewBox.Image = Nothing
			Else
				Dim ImgData() As Byte = TryCast(r(Pos).ItemArray(8), Byte())
				If ImgData Is Nothing Then
					PreviewBox.Image = Nothing
				Else
					Using ms As New MemoryStream(ImgData)
						PreviewBox.Image = Image.FromStream(ms)
					End Using
				End If
			End If
		End Sub

		Private Sub btnOpen_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnShowInExcel.Click
			If CurrentFilename Is Nothing Then
				MessageBox.Show("There is no open file")
				Return
			End If
			System.Diagnostics.Process.Start(CurrentFilename)
		End Sub

		Private Sub btnConvert_Click(ByVal sender As Object, ByVal e As System.EventArgs)
			'This is not yet implemented...
			Dim Pos As Integer = GetImagePos
			If Pos < 0 Then
				MessageBox.Show("There is no selected image", "Error")
				Return
			End If
			If CompressForm Is Nothing Then
				CompressForm = New TCompressForm()
			End If
			CompressForm.ImageToUse = CType(dataSet1.Tables("ImageDataTable").Rows(Pos).ItemArray(8), Byte())
			CompressForm.XlsFilename = CurrentFilename
			CompressForm.ShowDialog()
		End Sub

		Private Sub FilesListBox_DrawItem(ByVal sender As Object, ByVal e As System.Windows.Forms.DrawItemEventArgs) Handles FilesListBox.DrawItem
			If e.Index < 0 Then
				Return
			End If
			e.DrawBackground()
			Dim myBrush As Brush = Brushes.Black

			Dim ImageInfo As TImageInfo = CType(CType(sender, ListBox).Items(e.Index), TImageInfo)
			If Not ImageInfo.HasImages Then
				myBrush = Brushes.Silver
			End If
			If ImageInfo.HasCrop Then
				myBrush = Brushes.Red
			End If

			Dim NewStyle As FontStyle
			If ImageInfo.HasARGB Then
				NewStyle = FontStyle.Bold
			Else
				NewStyle = FontStyle.Regular
			End If
			Using MyFont As New Font(e.Font, NewStyle)
				e.Graphics.DrawString(ImageInfo.ToString(), MyFont, myBrush, e.Bounds, StringFormat.GenericDefault)
			End Using
			e.DrawFocusRectangle()
		End Sub

		Private Sub btnSaveImage_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSaveAsImage.Click
			Dim Pos As Integer = GetImagePos
			If Pos < 0 Then
				MessageBox.Show("There is no selected image to save", "Error")
				Return
			End If

			Dim ext As String = dataSet1.Tables("ImageDataTable").Rows(Pos).ItemArray(4).ToString().ToLower()
			saveImageDialog.DefaultExt = ext
			saveImageDialog.Filter = ext & " Images|*." & ext
			If saveImageDialog.ShowDialog() <> System.Windows.Forms.DialogResult.OK Then
				Return
			End If
			Dim ImgData() As Byte = CType(dataSet1.Tables("ImageDataTable").Rows(Pos).ItemArray(8), Byte())
			Using fs As New FileStream(saveImageDialog.FileName, FileMode.Create)
				fs.Write(ImgData, 0, ImgData.Length)
			End Using

		End Sub

		Private Sub btnInfo_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnInfo.Click
			MessageBox.Show("FlexCelImageExplorer is a small application targeted to reduce the size on images inside Excel files." & vbLf & "On the current version you can see the image properties and extract the images to disk.")

		End Sub

		Private Sub btnStretchPreview_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStretchPreview.Click
			If btnStretchPreview.Checked Then
				PreviewBox.SizeMode = PictureBoxSizeMode.StretchImage
			Else
				PreviewBox.SizeMode = PictureBoxSizeMode.Normal
			End If
		End Sub

	End Class

	Friend Class TImageInfo
		Friend File As FileInfo
		Friend IsValidFile As Boolean
		Friend HasCrop As Boolean
		Friend HasImages As Boolean
		Friend HasARGB As Boolean

		Public Sub New(ByVal aFile As FileInfo, ByVal aIsValidFile As Boolean, ByVal aHasCrop As Boolean, ByVal aHasImages As Boolean, ByVal aHasARGB As Boolean)
			File = aFile
			HasCrop = aHasCrop
			HasImages = aHasImages
			IsValidFile = aIsValidFile
			HasARGB = aHasARGB
		End Sub

		Public Overrides Function ToString() As String
			If Not IsValidFile Then
				Return " (*)" & File.ToString()
			End If
			Return File.ToString()
		End Function

	End Class
End Namespace

File: csharp/VS2019/netframework/Modules/10.API/30.FlexCelImageExplorer/MainForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.IO;

using FlexCel.Core;
using FlexCel.XlsAdapter;

namespace FlexCelImageExplorer
{
    /// <summary>
    /// Image Explorer.
    /// </summary>
    public partial class mainForm: System.Windows.Forms.Form
    {

        public mainForm()
        {
            InitializeComponent();
            GetCurrencyManager.CurrentChanged += new EventHandler(CurrentRowChanged);
            ResizeToolbar(mainToolbar);
        }

        private void ResizeToolbar(ToolStrip toolbar)
        {

            using (Graphics gr = CreateGraphics())
            {
                double xFactor = gr.DpiX / 96.0;
                double yFactor = gr.DpiY / 96.0;
                toolbar.ImageScalingSize = new Size((int)(24 * xFactor), (int)(24 * yFactor));
                toolbar.Width = 0; //force a recalc of the buttons.
            }
        }

        private string CurrentFilename = null;
        private TCompressForm CompressForm;

        private CurrencyManager GetCurrencyManager
        {
            get
            {
                return (CurrencyManager)this.BindingContext[dataGrid.DataSource, dataGrid.DataMember];
            }
        }

        private int GetImagePos
        {
            get
            {
                return GetCurrencyManager.Position;
            }
        }

        private void btnExit_Click(object sender, System.EventArgs e)
        {
            Close();
        }

        private void btnOpenFile_Click(object sender, System.EventArgs e)
        {
            if (openFileDialog1.ShowDialog() != DialogResult.OK) return;
            OpenFile(openFileDialog1.FileName);
            if (cbScanFolder.Checked) FillListBox();
        }

        private bool GetHasCrop(TImageProperties ImgProps)
        {
            return ImgProps.CropArea.CropFromLeft != 0 ||
                ImgProps.CropArea.CropFromRight != 0 ||
                ImgProps.CropArea.CropFromTop != 0 ||
                ImgProps.CropArea.CropFromBottom != 0;
        }

        private void FillListBox()
        {
            lblFolder.Text = "Files on folder: " + Path.GetDirectoryName(openFileDialog1.FileName);
            DirectoryInfo di = new DirectoryInfo(Path.GetDirectoryName(openFileDialog1.FileName));
            FileInfo[] Fi = di.GetFiles("*.xls");
            FilesListBox.Items.Clear();

            TImageInfo[] Files = new TImageInfo[Fi.Length];

            for (int k = 0; k < Fi.Length; k++)
            {
                FileInfo f = Fi[k];
                bool HasCrop = false;
                bool HasARGB = false;
                XlsFile x1 = new XlsFile();

                bool HasImages = false;

                try
                {
                    x1.Open(f.FullName);
                    for (int sheet = 1; sheet <= x1.SheetCount; sheet++)
                    {
                        x1.ActiveSheet = sheet;
                        for (int i = x1.ImageCount; i > 0; i--)
                        {
                            HasImages = true;
                            TImageProperties ip = x1.GetImageProperties(i);
                            if (!HasCrop) HasCrop = GetHasCrop(ip);

                            TXlsImgType imgType = TXlsImgType.Unknown;
                            using (MemoryStream ms = new MemoryStream())
                            {
                                x1.GetImage(i, ref imgType, ms);
                                FlexCel.Pdf.TPngInformation PngInfo = FlexCel.Pdf.TPdfPng.GetPngInfo(ms);
                                if (PngInfo != null)
                                {
                                    HasARGB = PngInfo.ColorType == 6;
                                }
                            }

                        }
                    }
                }
                catch (Exception)
                {
                    Files[k] = new TImageInfo(f, false, false, false, false);
                    continue;
                }

                Files[k] = new TImageInfo(f, true, HasCrop, HasImages, HasARGB);
            }

            FilesListBox.Items.AddRange(Files);
        }

        private void OpenFile(string FileName)
        {
            ImageDataTable.Rows.Clear();

            try
            {
                XlsFile Xls = new XlsFile(true);
                CurrentFilename = FileName;
                Xls.Open(FileName);

                for (int sheet = 1; sheet <= Xls.SheetCount; sheet++)
                {
                    Xls.ActiveSheet = sheet;
                    for (int i = Xls.ImageCount; i > 0; i--)
                    {
                        TXlsImgType ImageType = TXlsImgType.Unknown;
                        byte[] ImgBytes = Xls.GetImage(i, ref ImageType);
                        TImageProperties ImgProps = Xls.GetImageProperties(i);
                        object[] ImgData = new object[ImageDataTable.Columns.Count];
                        ImgData[0] = Xls.SheetName;
                        ImgData[1] = i;
                        ImgData[4] = ImageType.ToString();
                        ImgData[7] = Xls.GetImageName(i);
                        ImgData[8] = ImgBytes;
                        ImgData[9] = GetHasCrop(ImgProps);


                        using (MemoryStream ms = new MemoryStream(ImgBytes))
                        {
                            FlexCel.Pdf.TPngInformation PngInfo = FlexCel.Pdf.TPdfPng.GetPngInfo(ms);
                            if (PngInfo != null)
                            {
                                ImgData[2] = PngInfo.Width;
                                ImgData[3] = PngInfo.Height;
                                string s = String.Empty;
                                int bpp = 0;

                                if ((PngInfo.ColorType & 4) != 0)
                                {
                                    s += "ALPHA-";
                                    bpp = 1;
                                }
                                if ((PngInfo.ColorType & 2) == 0)
                                {
                                    s += "Grayscale -" + (1 << PngInfo.BitDepth).ToString() + " shades. ";
                                    bpp = 1;
                                }
                                else
                                {
                                    if ((PngInfo.ColorType & 1) == 0)
                                    {
                                        bpp += 3;
                                        s += "RGB - " + (PngInfo.BitDepth * (bpp)).ToString() + "bpp.  ";
                                    }
                                    else
                                    {
                                        s += "Indexed - " + (1 << PngInfo.BitDepth).ToString() + " colors. ";
                                        bpp = 1;
                                    }
                                }

                                ImgData[5] = s;

                                ImgData[6] = (Math.Round(PngInfo.Width * PngInfo.Height * PngInfo.BitDepth * bpp / 8f / 1024f)).ToString() + " kb.";
                            }
                            else
                            {
                                ms.Position = 0;
                                try
                                {
                                    using (Image Img = Image.FromStream(ms))
                                    {
                                        Bitmap Bmp = Img as Bitmap;
                                        if (Bmp != null)
                                        {
                                            ImgData[5] = Bmp.PixelFormat.ToString() + "bpp";
                                        }
                                        ImgData[2] = Img.Width;
                                        ImgData[3] = Img.Height;
                                    }
                                }
                                catch (Exception)
                                {
                                    ImgData[2] = -1;
                                    ImgData[3] = -1;
                                    ImgData[5] = null;
                                    ImgData[8] = null;

                                }
                            }
                        }


                        ImageDataTable.Rows.Add(ImgData);
                    }
                }

            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Error");
                dataGrid.CaptionText = "No file selected";
                CurrentFilename = null;
                return;
            }
            dataGrid.CaptionText = "Selected file: " + FileName;
            CurrentRowChanged(GetCurrencyManager, null);
        }

        private void FilesListBox_SelectedIndexChanged(object sender, System.EventArgs e)
        {
            TImageInfo ImageInfo = (TImageInfo)FilesListBox.SelectedItem;
            if (ImageInfo == null) return;
            OpenFile(ImageInfo.File.FullName);
        }

        public void CurrentRowChanged(object sender, System.EventArgs e)
        {
            int Pos = ((BindingManagerBase)sender).Position;
            if (Pos < 0)
            {
                PreviewBox.Image = null;
                return;
            }

            DataRowCollection r = dataSet1.Tables["ImageDataTable"].Rows;
            if (Pos < 0 || Pos >= r.Count)
            {
                PreviewBox.Image = null;
            }
            else
            {
                byte[] ImgData = r[Pos].ItemArray[8] as byte[];
                if (ImgData == null)
                {
                    PreviewBox.Image = null;
                }
                else
                {
                    using (MemoryStream ms = new MemoryStream(ImgData))
                    {
                        PreviewBox.Image = Image.FromStream(ms);
                    }
                }
            }
        }

        private void btnOpen_Click(object sender, System.EventArgs e)
        {
            if (CurrentFilename == null)
            {
                MessageBox.Show("There is no open file");
                return;
            }
            System.Diagnostics.Process.Start(CurrentFilename);
        }

        private void btnConvert_Click(object sender, System.EventArgs e)
        {
            //This is not yet implemented...
            int Pos = GetImagePos;
            if (Pos < 0)
            {
                MessageBox.Show("There is no selected image", "Error");
                return;
            }
            if (CompressForm == null) CompressForm = new TCompressForm();
            CompressForm.ImageToUse = (byte[])dataSet1.Tables["ImageDataTable"].Rows[Pos].ItemArray[8];
            CompressForm.XlsFilename = CurrentFilename;
            CompressForm.ShowDialog();
        }

        private void FilesListBox_DrawItem(object sender, System.Windows.Forms.DrawItemEventArgs e)
        {
            if (e.Index < 0) return;
            e.DrawBackground();
            Brush myBrush = Brushes.Black;

            TImageInfo ImageInfo = (TImageInfo)((ListBox)sender).Items[e.Index];
            if (!ImageInfo.HasImages)
            {
                myBrush = Brushes.Silver;
            }
            if (ImageInfo.HasCrop)
            {
                myBrush = Brushes.Red;
            }

            FontStyle NewStyle;
            if (ImageInfo.HasARGB) NewStyle = FontStyle.Bold; else NewStyle = FontStyle.Regular;
            using (Font MyFont = new Font(e.Font, NewStyle))
            {
                e.Graphics.DrawString(ImageInfo.ToString(),
                    MyFont, myBrush, e.Bounds, StringFormat.GenericDefault);
            }
            e.DrawFocusRectangle();
        }

        private void btnSaveImage_Click(object sender, System.EventArgs e)
        {
            int Pos = GetImagePos;
            if (Pos < 0)
            {
                MessageBox.Show("There is no selected image to save", "Error");
                return;
            }

            string ext = dataSet1.Tables["ImageDataTable"].Rows[Pos].ItemArray[4].ToString().ToLower();
            saveImageDialog.DefaultExt = ext;
            saveImageDialog.Filter = ext + " Images|*." + ext;
            if (saveImageDialog.ShowDialog() != DialogResult.OK) return;
            byte[] ImgData = (byte[])dataSet1.Tables["ImageDataTable"].Rows[Pos].ItemArray[8];
            using (FileStream fs = new FileStream(saveImageDialog.FileName, FileMode.Create))
            {
                fs.Write(ImgData, 0, ImgData.Length);
            }

        }

        private void btnInfo_Click(object sender, System.EventArgs e)
        {
            MessageBox.Show("FlexCelImageExplorer is a small application targeted to reduce the size on images inside Excel files.\n" +
                "On the current version you can see the image properties and extract the images to disk.");

        }

        private void btnStretchPreview_Click(object sender, EventArgs e)
        {
            if (btnStretchPreview.Checked)
                PreviewBox.SizeMode = PictureBoxSizeMode.StretchImage;
            else
                PreviewBox.SizeMode = PictureBoxSizeMode.Normal;
        }

    }

    class TImageInfo
    {
        internal FileInfo File;
        internal bool IsValidFile;
        internal bool HasCrop;
        internal bool HasImages;
        internal bool HasARGB;

        public TImageInfo(FileInfo aFile, bool aIsValidFile, bool aHasCrop, bool aHasImages, bool aHasARGB)
        {
            File = aFile;
            HasCrop = aHasCrop;
            HasImages = aHasImages;
            IsValidFile = aIsValidFile;
            HasARGB = aHasARGB;
        }

        public override string ToString()
        {
            if (!IsValidFile)
                return " (*)" + File.ToString();
            return File.ToString();
        }

    }
}
